dotcom-rendering/src/model/enhance-H3s.ts

```diff
@@ -0,0 +1,92 @@
+import { JSDOM } from 'jsdom';
+
+const isFalseH3 = (element: CAPIElement): boolean => {
+	if (!element) return false;
+	// Checks if this element is a 'false h3' based on the convention: <p><strong><H3 text</strong></p>
+	if (
+		element._type !== 'model.dotcomrendering.pageElements.TextBlockElement'
+	) {
+		return false;
+	}
+	const frag = JSDOM.fragment(element.html);
+	if (!frag || !frag.firstElementChild) return false;
+	const html = frag.firstElementChild.outerHTML;
+	// The following things must be true for an element to be a faux H3
+	const hasPwrapper = frag.firstElementChild.nodeName === 'P';
+	const containsStrongtags =
+		frag.firstElementChild.outerHTML.includes('<strong>');
+	const doesNotContainLinks =
+		!frag.firstElementChild.outerHTML.includes('<a>');
+	const startStrong = html.startsWith('<p><strong>');
+	const endsStrong = html.endsWith('</strong></p>');
+
+	return (
+		hasPwrapper &&
+		containsStrongtags &&
+		doesNotContainLinks &&
+		startStrong &&
+		endsStrong
+	);
+};
+
+const extractH3 = (element: CAPIElement): string => {
+	// Extract the text based on the convention: <p><strong><H3 text</strong></p>
+	const textElement = element as TextBlockElement;
+	const frag = JSDOM.fragment(textElement.html);
+	if (isFalseH3(element)) {
+		return (
+			frag.firstElementChild?.innerHTML
+				.split('<strong>')
+				.join('')
+				.split('</strong>')
+				.join('') ?? ''
+		);
+	}
+	return '';
+};
+
+const enhance = (elements: CAPIElement[]): CAPIElement[] => {
+	/**
+	 * Why not just add H3s in Composer?
+	 * Truth is, you can't. So to get around this there's a convention that says if
+	 * you insert <p><strong>Faux H3!</strong>,</p> then we replace it with a h3 tag
+	 * instead.
+	 */
+	const withH3s: CAPIElement[] = [];
+	elements.forEach((thisElement) => {
```

Comment:
I know this is being used elsewhere, but I think it's an anti-pattern.

A `map` is much more explicit and consistent. It should be trivial to refactor.

Reply:
I don't think it's entirely fair to call `forEach` an anti-pattern 🤣 . It's a valid feature of the javascript language

Reply:
`forEach` has its uses. It's a function that invokes side-effects for every item in an array.

In this case, however, these side effects have the property of mutating an array which ends up being the exact same length as the input array. There's a more explicit method that does exactly that: `map`. We generally favour being explicit and try to follow the principle of least surprise.

So, for this reason, using `forEach` to perform imperatively the exact same thing as `map` is “a bit of an anti-pattern”. If you wanted an array of a different length than the original array, or if you wanted to register listeners, or log things to the console, I would probably still recommend a `for…of` loop over `forEach` as it becomes more explicit that you will have side effects.

Reply:
For comparison, the resulting code is nearly identical, but you also get the added benefit of the TypeScript compiler catching the “passthrough“ case if you had forgotten to push it.

```ts
const withH3s: CAPIElement[] = elements.map((thisElement) => {
	if (isTextBlockElement(thisElement) && isFalseH3(thisElement)) {
		const h3Text = extractH3(thisElement);

		return {
			...thisElement,
			html: `<h3>${h3Text}</h3>`,
		};
	} else {
		// Pass through
		return thisElement;
	}
});
```

Reply:
I think map is a perfectly fine alternative here for the reasons you've given, and thanks for the example, that's useful!

I also think forEach is a valid solution; it's in use in lots of places and has not caused any problems. The term 'anti-pattern' has negative connotations which I don't think are fair given the differences here are ones of paradigm (functional programming vs. imperative) which are subjects of debate rather than concrete facts.

My personal preference these days is to use `map` and I generally think the community is moving this way but I am happy to work on code that uses `forEach` and tend to leave these sorts of decisions up to developers.

Reply:
I was unaware of the connotation of anti-pattern. I was trying to indicate that in this case there is an alternative that is more explicit, and therefore [better aligned to the guardian’s “coding with empathy” recommendations](https://github.com/guardian/recommendations/blob/main/coding-with-empathy.md#recommendations).

Reply:
Thank you both! Will change this to a map as it makes sense in this case.